### ISSUE TITLE
Bug: Failed to resolve: ai.coqui:libstt:0.9.3

### ISSUE THREAD

**Issue by soudalan** [coqui-ai/STT#1844]
After checking out the project and opening it in Android Studio (4.1), running the app throws this error:

Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
Could not find ai.coqui:libstt:0.9.3.
Searched in the following locations:
- https://dl.google.com/dl/android/maven2/ai/coqui/libstt/0.9.3/libstt-0.9.3.pom
- https://jcenter.bintray.com/ai/coqui/libstt/0.9.3/libstt-0.9.3.pom

**Comment by reuben** [comment]
Hey @soudalan. That's right, currently we only have pre-built packages published for Python on Linux, macOS and Windows. We're working to bring the rest of the Continuous Integration system up to speed. Windows support has just been merged today, in fact. I'll close this bug once the Java builds are running and the package is available.

**Comment by soudalan** [comment]
Thanks for clarifying, looking forward to it! I assume in the meantime, using the old DeepSpeechDemo is my only alternative for Android then?

**Comment by reuben** [comment]
That, or building from source.

**Comment by soudalan** [comment]
How would I build from the source? Not sure if there was a wiki or page anywhere explaining the process.

**Comment by reuben** [comment]
Following the [building docs](https://stt.readthedocs.io/en/latest/BUILDING.html) until the [configure step](https://stt.readthedocs.io/en/latest/BUILDING.html#tensorflow-configure-with-bazel), then switch to the [Android build steps](https://stt.readthedocs.io/en/latest/BUILDING.html#building-libstt-so-for-android) which show how to build libstt.so and then libstt.aar. But as I'm taking a look now, I'm realizing the state of building from source is not ideal. You'll need to manually copy the built libstt.so into the correct `libstt/libs` folder, and then also edit [native_client/java/libstt/gradle.properties](https://github.com/coqui-ai/STT/blob/main/native_client/java/libstt/gradle.properties) to keep only the architecture you're interested in using, before you run the `./gradlew libstt:build` step.

Hope this helps. If you try it and run into any problems please let me know so I can improve the docs. I'll also file an issue for improving the docs and the build system here to make it easier to build from source.

**Comment by soudalan** [comment]
I forgot to ask this earlier, but it seems I will need a GPU to build from the source? And assuming I don't have a GPU available, are there any other alternatives?

**Comment by reuben** [comment]
No, you don't need a GPU to build from source. What made you think that?

**Comment by soudalan** [comment]
Got it. The setup for Tensorflow asked for a path to my CUDA setup, so wasn't sure if I would also need a GPU prepared as well.

**Comment by reuben** [comment]
Ah, yes. You need CUDA installed to make GPU builds, but no actual GPUs are required. And you can also just not enable CUDA in the configure step to do CPU only builds.

**Comment by soudalan** [comment]
Under the "Building libstt.so for Android" section, when I run the command
`bazel build --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic --config=android --config=android_arm --define=runtime=tflite --action_env ANDROID_NDK_API_LEVEL=21 --cxxopt=-std=c++14 --copt=-D_GLIBCXX_USE_C99 //native_client:libstt.so`
, I get the error:

> ERROR: Skipping '//native_client:libstt.so': no such package 'native_client': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.

What BUILD file do I need to add exactly?

**Comment by soudalan** [comment]
Under the "Building libstt.so for Android" section, when I run the command
```bazel build --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic --config=android --config=android_arm --define=runtime=tflite --action_env ANDROID_NDK_API_LEVEL=21 --cxxopt=-std=c++14 --copt=-D_GLIBCXX_USE_C99 //native_client:libstt.so```
, I get the error:

> ERROR: Skipping '//native_client:libstt.so': no such package 'native_client': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.

This is after the symlink command `ln -s ../native_client`.
What BUILD file do I need to add exactly?

**Comment by reuben** [comment]
If you created the symlink inside the `tensorflow` folder and if you're running `bazel build` from the `tensorflow` folder, that's all that should be needed.

**Comment by soudalan** [comment]
Hmm, that's what I just did but it's throwing the error still.

**Comment by reuben** [comment]
Are you on Windows by any chance?

**Comment by soudalan** [comment]
No, I'm on Mac OS (Big Sur)

**Comment by soudalan** [comment]
> ERROR: Skipping '//native_client:libstt.so': no such package 'native_client': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /Users/jlam/Documents/coqui/tensorflow/native_client
WARNING: Target pattern parsing failed.
ERROR: no such package 'native_client': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /Users/jlam/Documents/coqui/tensorflow/native_client
INFO: Elapsed time: 0.156s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
jlam-macbookpro% ls
ACKNOWLEDGMENTS		LICENSE			configure.py
AUTHORS			README.md		models.BUILD
BUILD			RELEASE.md		native_client
CODEOWNERS		SECURITY.md		tensorflow
CODE_OF_CONDUCT.md	WORKSPACE		third_party
CONTRIBUTING.md		arm_compiler.BUILD	tools
ISSUES.md		configure
ISSUE_TEMPLATE.md	configure.cmd
jlam-macbookpro% pwd
/Users/jlam/Documents/coqui/tensorflow
jlam-macbookpro%

**Comment by reuben** [comment]
Can you run, from the `tensorflow` folder, the following command and post their outputs?
`ls -lha`
`ls -lha native_client`

**Comment by soudalan** [comment]
```
jlam-macbookpro% ls -lha
total 1112
drwxr-xr-x  31 jlam  primarygroup   992B Apr 20 11:37 .
drwxr-xr-x   4 jlam  primarygroup   128B Apr 19 15:23 ..
-rw-r--r--   1 jlam  primarygroup    29K Apr 19 15:26 .bazelrc
-rw-r--r--   1 jlam  primarygroup     6B Apr 19 15:26 .bazelversion
drwxr-xr-x  13 jlam  primarygroup   416B Apr 19 15:27 .git
drwxr-xr-x   6 jlam  primarygroup   192B Apr 19 15:26 .github
-rw-r--r--   1 jlam  primarygroup   896B Apr 19 15:26 .gitignore
lrwxr-xr-x   1 jlam  primarygroup    34B Apr 19 15:26 .pylintrc -> tensorflow/tools/ci_build/pylintrc
-rw-r--r--   1 jlam  primarygroup   688B Apr 20 08:55 .tf_configure.bazelrc
-rw-r--r--   1 jlam  primarygroup   2.2K Apr 19 15:26 ACKNOWLEDGMENTS
-rw-r--r--   1 jlam  primarygroup   349B Apr 19 15:26 AUTHORS
-rw-r--r--   1 jlam  primarygroup    95B Apr 19 15:26 BUILD
-rw-r--r--   1 jlam  primarygroup   656B Apr 19 15:26 CODEOWNERS
-rw-r--r--   1 jlam  primarygroup   5.2K Apr 19 15:26 CODE_OF_CONDUCT.md
-rw-r--r--   1 jlam  primarygroup   9.7K Apr 19 15:26 CONTRIBUTING.md
-rw-r--r--   1 jlam  primarygroup   606B Apr 19 15:26 ISSUES.md
-rw-r--r--   1 jlam  primarygroup   2.2K Apr 19 15:26 ISSUE_TEMPLATE.md
-rw-r--r--   1 jlam  primarygroup    13K Apr 19 15:26 LICENSE
-rw-r--r--   1 jlam  primarygroup    21K Apr 19 15:26 README.md
-rw-r--r--   1 jlam  primarygroup   333K Apr 19 15:26 RELEASE.md
-rw-r--r--   1 jlam  primarygroup    13K Apr 19 15:26 SECURITY.md
-rw-r--r--   1 jlam  primarygroup   665B Apr 19 15:26 WORKSPACE
-rw-r--r--   1 jlam  primarygroup   1.1K Apr 19 15:26 arm_compiler.BUILD
-rwxr-xr-x   1 jlam  primarygroup   285B Apr 19 15:26 configure
-rw-r--r--   1 jlam  primarygroup   782B Apr 19 15:26 configure.cmd
-rw-r--r--   1 jlam  primarygroup    52K Apr 19 15:26 configure.py
-rw-r--r--   1 jlam  primarygroup   328B Apr 19 15:26 models.BUILD
lrwxr-xr-x   1 jlam  primarygroup    16B Apr 20 11:37 native_client -> ../native_client
drwxr-xr-x  36 jlam  primarygroup   1.1K Apr 20 11:35 tensorflow
drwxr-xr-x  97 jlam  primarygroup   3.0K Apr 19 15:27 third_party
drwxr-xr-x   4 jlam  primarygroup   128B Apr 19 15:29 tools
jlam-macbookpro% ls -lha native_client
lrwxr-xr-x  1 jlam  primarygroup    16B Apr 20 11:37 native_client -> ../native_client
```

**Comment by reuben** [comment]
Yeah it looks exactly like it should. And can you also do `cat native_client/BUILD` from the `tensorflow` folder? To be clear, that symlink should be pointing to the `native_client` folder inside the STT repo.

**Comment by soudalan** [comment]
It says 

> cat: native_client/BUILD: No such file or directory

**Comment by reuben** [comment]
Yeah, so either you're not in the right `tensorflow` folder (the submodule) or something in your system is broken in a way I've never seen. Make sure you have the following structure:

```
STT:
  - native_client/
    - BUILD
    - ...
  - tensorflow/
    - native_client -> ../native_client
    - ...
  - ...
```

**Comment by soudalan** [comment]
Good catch! Turns out I had two separate Tensorflow subdirecties, and I was running these commands inside the one that wasn't part of the STT directory. Thanks for pointing that out!

**Comment by soudalan** [comment]
I ran into this issue when building specifically for Android:

> ERROR: /Users/jlam/Documents/STT/tensorflow/tensorflow/lite/kernels/internal/BUILD:687:1: C++ compilation of rule '//tensorflow/lite/kernels/internal:kernel_utils' failed (Exit 1)
src/main/tools/process-wrapper-legacy.cc:58: "execvp(/bin/false, ...)": No such file or directory
Target //native_client:libstt.so failed to build

This is after running
```
bazel build --workspace_status_command="bash native_client/bazel_workspace_status_cmd.sh" --config=monolithic --config=android --config=android_arm --define=runtime=tflite --action_env ANDROID_NDK_API_LEVEL=21 --cxxopt=-std=c++14 --copt=-D_GLIBCXX_USE_C99 //native_client:libstt.so
```

**Comment by reuben** [comment]
Did you try the second solution posted in that issue? https://github.com/tensorflow/tensorflow/issues/2640#issuecomment-223701349

**Comment by soudalan** [comment]
As this was posted 5 years ago, both versions are rather outdated now, so I wasn't sure if it made sense to build with such old versions. It could possibly still be an issue with version mis-matches, so I will try changing the version.

**Comment by NadeemAraman** [comment]
Please i need a firm answer , do i need to use Gpu (Cuda paths) to build libstt? If no i did get through all the steps but ended up with this error , any help would be very appreciated 🙏

`function 'virtual Metadata* ModelState::decode_metadata(const DecoderState&, size_t)':
native_client/modelstate.cc:73:3: error: too many initializers for 'Metadata'
   };
   ^`
``
